### PR TITLE
Cloudchamber commands

### PR DESCRIPTION
**Pull request by gabivlj**
This is a PR that adds 'cloudchamber' commands.

# Notes
- They don't appear in wrangler --help.
- Each command is in its own commit for ease of review, however there is many commits so I think we can have an approach of approving commit one by one.
- The commits that change things in C3 are important. Below I'm going to provide the changes that happen in @cloudflare/cli that are relevant.

# Cloudflare CLI changes
- [CLI C3: processArguments in CLI and Chalk of wrangler in cli](https://github.com/cloudflare/workers-sdk/commit/3389e2f80ff5f7fdeca45cb6b3bf9be1055cb70a): We need processArguments in wrangler, seems sensible to do this.
- We are downgrading chalk to a version that wrangler uses in `@cloudflare/cli`, any other solution to this problem? C3 used to use 5.2.0, but it doesn't seem to work when I import to wrangler because wrangler is stuck on "^2.4.2". Do I upgrade wrangler or downgrade @cloudflare/cli chalk version?
- We're introducing a new component that is implemented by clack: MultiSelectPrompt. We try to introduce some kind of simple pagination that can be improved, but it's used as last resort.
- We're introducing a new component implemented by us: SelectRefreshablePrompt. This component introduces infinite scrolling and a way for the caller to refresh the items on 'R' press. This is important for the `cloudchamber list` view, as it provides a lot of interactivity. We tried very hard to not introduce on this view an ink dependency. We understand any code that you might want to see some improvements here, but we like the result and how it works.

# wrangler cloudchamber architecture
- We use a code-generated HTTP client from a OpenAPI yaml spec. If this is too much or this is not desirable, we can discuss it; however it provides fast iteration for us and we like it a lot due to the typesafety and introducing special error handling.
- Each command is on its own file, except images.ts that introduces `cloudchamber registries`.

# TODOs
- [x] Changesets 
- [x] Lints and fixes

### CODE REVIEW

File: packages/wrangler/src/index.ts
Comment (penalosa) [inline packages/wrangler/src/index.ts:379]: This leads to a suboptimal experience when running `wrangler cloudchamber` (no args). For most commands (e.g. `wrangler r2`) not specifying a subcommand will lead to help output e.g.

```
wrangler r2

📦 Interact with an R2 store

Commands:
  wrangler r2 object  Manage R2 objects
  wrangler r2 bucket  Manage R2 buckets
```

For `wrangler cloudchamber` it leads to help output but _also_ includes a non-actionable error:

```
✘ [ERROR] Not enough non-option arguments: got 0, need at least 1


wrangler cloudchamber
```

Could you integrate with `subHelp`?

File: packages/wrangler/src/cloudchamber/index.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/index.ts:37]: In the rest of Wrangler we generally don't use `demandCommand` and `showHelpOnFail` in this way—see `index.ts`

File: packages/wrangler/src/cloudchamber/common.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/common.ts:1]: I'm not fully sure what some of the typing code here is doing (will dive a bit more into it to understand), but I think a lot could potentially be replaced by `StrictYargsOptionsToInterface`, which is what we use across Wrangler to translate between a Yargs options function and the arguments for the handler

File: packages/wrangler/src/cloudchamber/delete.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/delete.ts:13]: We try not to use `Argv` directly from Yargs since it doesn't include top-level Wrangler flags. Instead, `CommonYargsArgv` from `yargs-types.ts` should be used where possible

File: packages/wrangler/src/cloudchamber/index.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/index.ts:31]: This is fairly nested, and would probably benefit from being individual `.command()` calls readability-wise. To maintain the consistent `--json` flag, it might be worth following a pattern similar to `wrangler ai` for each command's options (https://github.com/gabivlj/workers-sdk/blob/f6b9c89cdf97bc1c260427640e635eeffaddc173/packages/wrangler/src/ai/options.ts#L11-L17)

**Review by penalosa — COMMENTED** [review]
Some initial comments

File: packages/wrangler/src/user/user.ts
Comment (gabivlj) [reply packages/wrangler/src/user/user.ts:355]: The argument I've seen is that most users won't ever have to interact with cloudchamber in the first place. So requesting scopes for them is not necessary.

File: packages/wrangler/src/cloudchamber/common.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/common.ts:1]: Yep, that's great, we would love to use that. We were looking for something like that, thanks for the suggestion. We can remove inferYargs with it.

handleFailure sets up a common error handling for our API types and handles some common stuff like authentication for the commands. 
We like this method because we're able to separate the specific domain args of the command in the first param, and the common config like json and wrangler config in the other parameter. This allows us to have good type guarantees in the domain command and even be able to pass directly the arg object into api calls directly without any random stuff in the JavaScript object.

{code}
		(args) =>
			listDeploymentsYargs(args).positional("deploymentIdPrefix", {
				describe:
					"Optional deploymentId to filter deployments\nThis means that 'list' will only showcase deployments that contain this ID prefix",
				type: "string",
			}),
		(args) =>
			handleFailure<typeof args>(async (deploymentArgs, config) => {
// here deploymentArgs has correct parameters, and config has json and wrangler config
{\code} 

LMK if its overkill but I like it a lot that it separates cleanly the types

File: packages/wrangler/src/cloudchamber/delete.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/delete.ts:13]: Agreed! Will change.

File: packages/wrangler/src/cloudchamber/index.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/index.ts:31]: Yeah might take that method and put it on a common file.

File: packages/wrangler/src/cloudchamber/common.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/common.ts:1]: I ended up removing this as this use-case is something we had more in our internal fork, now it's not necessary.

**Comment by gabivlj** [comment]
Addressed @penalosa comments on the last commit as they affected all cloudchamber commands.

File: packages/wrangler/src/cloudchamber/index.ts
Comment (irvinebroque) [inline packages/wrangler/src/cloudchamber/index.ts:24]:
```suggestion
			"Create a new deployment",
```

File: packages/wrangler/src/cloudchamber/index.ts
Comment (irvinebroque) [inline packages/wrangler/src/cloudchamber/index.ts:36]:
```suggestion
			"Modify an existing deployment",
```

File: packages/wrangler/src/user/user.ts
Comment (penalosa) [reply packages/wrangler/src/user/user.ts:355]: That makes sense, I suppose—I think the counterargument that we've generally used in that case is that unconditionally requesting scopes means that running `wrangler login` is all that anyone needs to use _any_ Wrangler command, and there's no need to re-login for specific commands

File: packages/wrangler/src/cloudchamber/ssh/ssh.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/ssh/ssh.ts:53]:
```suggestion
 *  Tries to retrieve the SSH key if sshKeyPath is a path to a ssh key.
```

File: packages/wrangler/src/cloudchamber/ssh/ssh.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/ssh/ssh.ts:45]:
```suggestion
		.option("public-key", {
```

File: packages/wrangler/src/cloudchamber/ssh/ssh.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/ssh/ssh.ts:68]: This should use `logger.error()` (as should any other uses of `console.*`)

File: packages/wrangler/src/cloudchamber/ssh/ssh.ts
Comment (penalosa) [inline packages/wrangler/src/cloudchamber/ssh/ssh.ts:75]: This could probably do with a debug log as well

File: packages/wrangler/src/cloudchamber/ssh/ssh.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/ssh/ssh.ts:68]: Can we just do logger.log on places json only is expected?

**Review by penalosa — APPROVED** [review]
LGTM! 🎉

**Review by petebacondarwin — APPROVED** [review]
LGTM for C3 - didn't look at anything else.

File: packages/cli/index.ts
Comment (dario-piotrowicz) [inline packages/cli/index.ts:140]: what's the difference between this and `crash` above? couldn't `crash` be generalized instead of defining a new function?

File: packages/cli/index.ts
Comment (dario-piotrowicz) [inline packages/cli/index.ts:142]: not `stderr`?

File: packages/cli/index.ts
Comment (gabivlj) [reply packages/cli/index.ts:140]: It prints an "ERROR" prefix, and I feel like an error method can convey different messaging from crash

Comment (dario-piotrowicz) [reply packages/cli/index.ts:140]: I see, regarding the prefix again it could be set conditionally (although I would be tempted to just always include it)

___

I see your point in `crash` vs `error`, but I must say that I don't fully agree, `crash` is just a synonym for "unrecoverable error" isn't it?

Also it does to me look like `crash` is just used for displaying errors (+ exiting the process):
Screenshot 2023-11-30 at 13 07 33

___

If we for example take:
```ts
crash("Failed to create pages project. See output above.");
```
([source](https://github.com/cloudflare/workers-sdk/blob/f20b02a90cea851c8b8d050cfc34ec310e3ff8f0/packages/create-cloudflare/src/pages.ts#L222))
could you explain to me the difference between that and
```ts
error(
			"There has been an error while loading your deployments: \n " +
				err.message
		);
```
([source](https://github.com/cloudflare/workers-sdk/pull/4310/files#diff-4e6941af0591cf1d2252f8793c29c620c19f9f35548769f1ce39b53226cd347eR79-R82))
?

They seem to convey practically the same exact meaning/semantic to me (and I don't see the benefit in having one being called `error` and the other `crash`)

___

Another example:
```ts
crash("Pages project isn't ready yet. Please try deploying again later.");
```
([source](https://github.com/cloudflare/workers-sdk/blob/f20b02a90cea851c8b8d050cfc34ec310e3ff8f0/packages/create-cloudflare/src/pages.ts#L251))
vs
```ts
	error(
		"Couldn't retrieve a new deployment: " + getDeploymentError.message
	);
```
([source](https://github.com/cloudflare/workers-sdk/pull/4310/files#diff-47cec4445ac6c9fb7fba1c8cb8f273d48d999ad3e7928176a71bfcea5a6080d4R313-R315))

Both indicate that the result from some kind of deployment isn't retrievable.

could you explain why the first is a crash and the second is an error?

Comment (gabivlj) [reply packages/cli/index.ts:140]: Well, for us we just found it to differentiate the rendering style. We didn't find changing the crash function signature to expose an error prefix an appealing thing to do at first.

**Comment by irvinebroque** [comment]
@gabivlj @skepticfx — possible to make it such that:

1. When you run `wrangler login --scopes-list`, the Cloudchamber read and write scopes are shown as available?
2. It is possible to use the `--scopes`  flag to explicitly login with the Cloudchamber scopes? Ex: `wrangler login --scopes "cloudchamber:read", "cloudchamber:write"

Solving (2) would ensure that there is simple and easy way for people to login with the right scopes, without making it such that the cloudchamber scopes are requested for all users.

**Comment by benbuzbee** [comment]
Could  someone with permissions please re-queue the builds which generate the NPM artifact? The NPM artifact for this PR seems to have expired. @petebacondarwin or @dario-piotrowicz perhaps?

@

**Comment by gabivlj** [comment]
Hi @dario-piotrowicz, I've replaced all the error calls with crash and addressed the feedback. I also had to do a rebase so it would be nice if C3 folks could take another pass to this.

**Comment by dario-piotrowicz** [comment]
@gabivlj the changes in [fde9904](https://github.com/cloudflare/workers-sdk/pull/4310/commits/fde9904cec050a37545ef3db9cd11424ab91e853) do look good to me, thanks a bunch! 😄

The c3 e2e tests seem to be failing here though so we need to make sure that they're not getting broken (since all are failing it feels like there could actually be something in these changes that breaks them)

Also there are some merge conflicts, those are my bad as I slightly tweaked the `crash` function, sorry about that 🙇 (but they should hopefully be very simple to solve 🙂)

I'll give the PR another review today or tomorrow (although there's a bunch going on here :face_with_spiral_eyes: 😅)

I think we should also wait for @jculvey to give this a review as he's practically implemented most if not all the C3 cli bits 🙂

**Review by petebacondarwin — COMMENTED** [review]
Looking good!
A few drive by comments... haven't had time to review it all yet.

File: packages/wrangler/src/cloudchamber/client/core/ApiError.ts
Comment (gabivlj) [reply packages/wrangler/src/cloudchamber/client/core/ApiError.ts:1]: This code (the client folder) is codegenerated from our OpenAPI spec

**Comment by gabivlj** [comment]
> @gabivlj the changes in [fde9904](https://github.com/cloudflare/workers-sdk/pull/4310/commits/fde9904cec050a37545ef3db9cd11424ab91e853) do look good to me, thanks a bunch! 😄
> 
> The c3 e2e tests seem to be failing here though so we need to make sure that they're not getting broken (since all are failing it feels like there could actually be something in these changes that breaks them)
> 
> Also there are some merge conflicts, those are my bad as I slightly tweaked the `crash` function, sorry about that 🙇 (but they should hopefully be very simple to solve 🙂)
> 
> I'll give the PR another review today or tomorrow (although there's a bunch going on here 😵‍💫 😅)
> 
> I think we should also wait for @jculvey to give this a review as he's practically implemented most if not all the C3 cli bits 🙂

Seems like the issues are with running `wrangler whoami` within C3, which triggers a login and that fails... I'm going to take a look at what whoami spits

**Comment by dario-piotrowicz** [comment]
> > @gabivlj the changes in [fde9904](https://github.com/cloudflare/workers-sdk/pull/4310/commits/fde9904cec050a37545ef3db9cd11424ab91e853) do look good to me, thanks a bunch! 😄
> > The c3 e2e tests seem to be failing here though so we need to make sure that they're not getting broken (since all are failing it feels like there could actually be something in these changes that breaks them)
> > Also there are some merge conflicts, those are my bad as I slightly tweaked the `crash` function, sorry about that 🙇 (but they should hopefully be very simple to solve 🙂)
> > I'll give the PR another review today or tomorrow (although there's a bunch going on here 😵‍💫 😅)
> > I think we should also wait for @jculvey to give this a review as he's practically implemented most if not all the C3 cli bits 🙂
> 
> Seems like the issues are with running `wrangler whoami` within C3, which triggers a login and that fails... I'm going to take a look at what whoami spits

@gabivlj, I'm planning to have C3 rely on the `--json` option discussed in https://github.com/cloudflare/workers-sdk/issues/4596 and C3 should rely on that one, so if you find issues here in regard on how C3 uses `wrangler whoami` keep in mind that it will likely get improved soon 🙂 (I'm mentioning this just so say that if you need to fix things regarding whoami I would suggest to add temporary workarounds and not stress over things too much here)

**Review by dario-piotrowicz — APPROVED** [review]
Just a few comments 🙂 

The C3 changes make sense to me so I'm giving a tenative approval on that side of things, although I didn't get the change to review everything here in details as there's a bunch to unpack here and I also lack context on the cloudchamber changes

**Comment by gabivlj** [comment]
> > > @gabivlj the changes in [fde9904](https://github.com/cloudflare/workers-sdk/pull/4310/commits/fde9904cec050a37545ef3db9cd11424ab91e853) do look good to me, thanks a bunch! 😄
> > > The c3 e2e tests seem to be failing here though so we need to make sure that they're not getting broken (since all are failing it feels like there could actually be something in these changes that breaks them)
> > > Also there are some merge conflicts, those are my bad as I slightly tweaked the `crash` function, sorry about that 🙇 (but they should hopefully be very simple to solve 🙂)
> > > I'll give the PR another review today or tomorrow (although there's a bunch going on here 😵‍💫 😅)
> > > I think we should also wait for @jculvey to give this a review as he's practically implemented most if not all the C3 cli bits 🙂
> > 
> > 
> > Seems like the issues are with running `wrangler whoami` within C3, which triggers a login and that fails... I'm going to take a look at what whoami spits
> 
> @gabivlj, I'm planning to have C3 rely on the `--json` option discussed in #4596 and C3 should rely on that one, so if you find issues here in regard on how C3 uses `wrangler whoami` keep in mind that it will likely get improved soon 🙂 (I'm mentioning this just so say that if you need to fix things regarding whoami I would suggest to add temporary workarounds and not stress over things too much here)

The debug log files apparently showcase that wrangler whoami is returning that the user is not logged in, then making C3 run wrangler login. I'm a bit confused on why is whoami failing to authenticate still, as I couldn't repro yet locally on why is this behavior :thinking:.

File: packages/cli/interactive.ts
Comment (gabivlj) [reply packages/cli/interactive.ts:390]: I agree. I think this was done to add some vertical margin, good call here as I completely forgot to comment it here :+1:

File: packages/cli/interactive.ts
Comment (Skye-31) [reply packages/cli/interactive.ts:411]: You could also go for
```suggestion
		for (const i in pages) {
			const pageIter = pages[i];
```

**Comment by gabivlj** [comment]
All tests seem to pass @dario-piotrowicz, I will address your comments now and let's see if we can merge this soon if everyone approves

File: packages/cli/interactive.ts
Comment (gabivlj) [reply packages/cli/interactive.ts:411]: Personally I'd go for the `let i = 0` because intent is clear enough for me

**Pull request merged**
